File: packages/plugin-contract/src/manifest.ts
export type PluginKind = "device-driver" | "integration" | "processing";
export type PluginRuntime = "in-process-ts" | "python-sidecar";

/**
 * Modelo de permisos deny-by-default (ADR-008, P8): lo que un plugin
 * declara que necesita tocar. Capabilities no viven acá a propósito — ya
 * son dinámicas por dispositivo vía `KanDeviceDriverPlugin.getCapabilities()`;
 * duplicarlas en el manifest sería la misma divergencia manifest/
 * implementación que docs/04 §5 pide evitar.
 */
export interface PluginPermissions {
  /** Kinds de dispositivo que este plugin necesita controlar. */
  devices: string[];
  network: boolean;
  /** Declarativo, ej. "read:user-uploads" — sin enforcement real todavía. */
  filesystem: string[];
}

export interface PluginManifest {
  id: string;
  version: string;
  displayName: string;
  kind: PluginKind;
  runtime: PluginRuntime;
  permissions: PluginPermissions;
  /** Placeholder para firma criptográfica real (docs/09, Año 1/2) — no verificado todavía. */
  signature?: string;
}

File: packages/kan-plugin-sdk-py/src/kan_plugin_sdk_py/manifest.py
"""Espejo a mano de packages/plugin-contract/src/manifest.ts.

`signature` sigue siendo un placeholder sin verificación real (mismo
límite documentado del lado TS) — no se firma nada acá todavía.
"""

from dataclasses import dataclass, field
from typing import Any, Literal, Optional

PluginKind = Literal["device-driver", "integration", "processing"]
PluginRuntime = Literal["in-process-ts", "python-sidecar"]


@dataclass
class PluginPermissions:
    devices: list[str] = field(default_factory=list)
    network: bool = False
    filesystem: list[str] = field(default_factory=list)

    def to_wire(self) -> dict[str, Any]:
        return {"devices": self.devices, "network": self.network, "filesystem": self.filesystem}


@dataclass
class PluginManifest:
    id: str
    version: str
    display_name: str
    kind: PluginKind
    runtime: PluginRuntime
    permissions: PluginPermissions
    signature: Optional[str] = None

    def to_wire(self) -> dict[str, Any]:
        wire: dict[str, Any] = {
            "id": self.id,
            "version": self.version,
            "displayName": self.display_name,
            "kind": self.kind,
            "runtime": self.runtime,
            "permissions": self.permissions.to_wire(),
        }
        if self.signature is not None:
            wire["signature"] = self.signature
        return wire

    @staticmethod
    def from_wire(data: dict[str, Any]) -> "PluginManifest":
        permissions_data = data.get("permissions", {})
        return PluginManifest(
            id=data["id"],
            version=data["version"],
            display_name=data["displayName"],
            kind=data["kind"],
            runtime=data["runtime"],
            permissions=PluginPermissions(
                devices=list(permissions_data.get("devices", [])),
                network=bool(permissions_data.get("network", False)),
                filesystem=list(permissions_data.get("filesystem", [])),
            ),
            signature=data.get("signature"),
        )
